coherence-distributed-configuration/src/main/java/io/micronaut/coherence/discovery/CoherenceConfigurationClient.java

```diff
@@ -0,0 +1,182 @@
+import org.reactivestreams.Publisher;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Singleton;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+@Singleton
+@BootstrapContextCompatible
+public class CoherenceConfigurationClient implements ConfigurationClient {
+
+    private static final Logger LOG = LoggerFactory.getLogger(CoherenceConfigurationClient.class);
+    private static final String DEFAULT_APPLICATION = "application";
+    /**
+     * The name of the map that is used to store configuration properties.
+     */
+    private static final String CONFIG_CACHE_NAME = "config-cache";
+
+    private final ApplicationConfiguration applicationConfiguration;
+    private final List<Flowable<PropertySource>> propertySources = new ArrayList<>();
+    private final CoherenceClientConfiguration coherenceClientConfiguration;
+
+    public CoherenceConfigurationClient(ApplicationConfiguration applicationConfiguration,
+                                        CoherenceClientConfiguration coherenceClientConfiguration) {
+        this.applicationConfiguration = applicationConfiguration;
+        this.coherenceClientConfiguration = coherenceClientConfiguration;
+        LOG.debug("Coherence Client configuration: " + coherenceClientConfiguration);
+    }
+
+    @Override
+    public Publisher<PropertySource> getPropertySources(Environment environment) {
+        ValueExtractor<String, String> keyExtractor = new KeyExtractor<>();
+        NamedCache<String, Object> configCache = getCache(CONFIG_CACHE_NAME, coherenceClientConfiguration);
+
+        Map<Integer, String> keys = buildKeys(applicationConfiguration, environment);
+        for (Map.Entry<Integer, String> entry : keys.entrySet()) {
+            final Integer priority = entry.getKey();
+            final String propertySource = entry.getValue();
+            final String propertySourcePath = propertySource + "/";
+            Remote.Predicate<String> pathPredicate = cacheKey -> cacheKey.startsWith(propertySourcePath) &&
+                    cacheKey.length() > propertySourcePath.length() &&
+                    !cacheKey.substring(propertySourcePath.length()).contains("/");
+            Filter<String> filter = Filters.predicate(keyExtractor, pathPredicate);
+            Map<String, Object> props = configCache.entrySet(filter)
+                    .stream()
+                    .collect(Collectors.toMap((Converter<Map.Entry<String, Object>, String>)
+                                    cacheEntry -> cacheEntry.getKey().substring(propertySourcePath.length()),
+                            Map.Entry::getValue));
+            Flowable<PropertySource> propertySourceFlowable = Flowable.just(PropertySource.of(propertySource, props, priority));
+            propertySources.add(propertySourceFlowable);
+        }
```

Comment:
I'm not quite sure I understand this logic, but it seems to me it would be *much* simpler to use multiple caches (cache per application/environment combination), and define property source for each, than to use a single cache and do all of this by mangling and querying config key names.

Let's discuss...